File: tree/104_maximum_depth_of_binary_tree.cc
#include <algorithm>
#include <queue>
#include <vector>

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

// 2023/12/21: 可以不看答案自己写出来递归.
int MaxDepth(TreeNode* root) {
  if (root == nullptr) {
    return 0;
  }
  return std::max(MaxDepth(root->left), MaxDepth(root->right)) + 1;
}

int MaxDepth(TreeNode* root) {
  if (root == nullptr) {
    return 0;
  }
  std::queue<TreeNode*> q;
  q.push(root);
  int ans = 0;
  while (!q.empty()) {
    ++ans;
    int count = q.size(); // 维护一个size
    while (count > 0) {
      TreeNode* tree_node = q.front();
      q.pop();
      if (tree_node->left) {
        q.push(tree_node->left);
      }
      if (tree_node->right) {
        q.push(tree_node->right);
      }
      --count;
    }
  }

  return ans;
}

File: tree/104_maximum_depth_of_binary_tree.py
from collections import deque
from typing import Optional
from tree.tree_node import TreeNode, create_tree_node


# 2026/05/18: 自己写出递归.
def max_depth(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0

    return max(max_depth(root.left), max_depth(root.right)) + 1


# 2026/05/18: 自己写出循环, Python 中把 deque 当作 queue 使用.
def max_depth2(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0

    dq = deque()
    dq.append(root)
    depth = 0
    while len(dq) != 0:
        depth += 1
        count = len(dq)
        for _ in range(count):
            curr = dq.popleft()
            if curr.left is not None:
                dq.append(curr.left)
            if curr.right is not None:
                dq.append(curr.right)

    return depth


root = [3, 9, 20, None, None, 15, 7]
root = create_tree_node(root)
print(max_depth2(root))
